bundler/spec/native_helpers/functions/version_resolver_spec.rb

```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+require "native_spec_helper"
+require "shared_contexts"
+
+require "dependabot/dependency"
+
+RSpec.describe Functions::VersionResolver do
+  include_context "in a temporary bundler directory"
+  include_context "stub rubygems compact index"
+
+  let(:version_resolver) do
+    described_class.new(
+      dependency_name: dependency_name,
+      dependency_requirements: dependency_requirements,
+      gemfile_name: gemfile_name,
+      lockfile_name: lockfile_name
+    )
+  end
+
+  let(:dependency_name) { "business" }
+  let(:dependency_requirements) do
+    [{
+      file: "Gemfile",
+      requirement: requirement_string,
+      groups: [],
+      source: source
+    }]
+  end
+  let(:source) { nil }
+
+  let(:rubygems_url) { "https://index.rubygems.org/api/v1/" }
+  let(:old_index_url) { rubygems_url + "dependencies" }
+
+  describe "#version_details" do
+    subject do
+      in_tmp_folder { version_resolver.version_details }
+    end
+
+    let(:gemfile_fixture_name) { "gemfile" }
+    let(:lockfile_fixture_name) { "gemfile.lock" }
+    let(:requirement_string) { " >= 0" }
+
+    its([:version]) { is_expected.to eq(Gem::Version.new("1.4.0")) }
+    its([:fetcher]) { is_expected.to eq("Bundler::Fetcher::CompactIndex") }
+
+    context "with a private gemserver source" do
+      include_context "stub rubygems compact index"
+
+      let(:gemfile_fixture_name) { "specified_source" }
+      let(:lockfile_fixture_name) { "specified_source.lock" }
+      let(:requirement_string) { ">= 0" }
+
+      before do
+        gemfury_url = "https://repo.fury.io/greysteil/"
+        gemfury_deps_url = gemfury_url + "api/v1/dependencies"
+
+        stub_request(:get, gemfury_url + "versions").
+          to_return(status: 200, body: fixture("ruby", "gemfury-index"))
+        stub_request(:get, gemfury_url + "info/business").to_return(status: 404)
+        stub_request(:get, gemfury_deps_url).to_return(status: 200)
+        stub_request(:get, gemfury_deps_url + "?gems=business,statesman").
+          to_return(status: 200, body: fixture("ruby", "gemfury_response"))
+        stub_request(:get, gemfury_deps_url + "?gems=business").
+          to_return(status: 200, body: fixture("ruby", "gemfury_response"))
+        stub_request(:get, gemfury_deps_url + "?gems=statesman").
+          to_return(status: 200, body: fixture("ruby", "gemfury_response"))
+      end
+
+      its([:version]) { is_expected.to eq(Gem::Version.new("1.9.0")) }
+      its([:fetcher]) { is_expected.to eq("Bundler::Fetcher::Dependency") }
```

Comment:
One thing I've realised in writing this test, if we get a version back from a _private_ registry, the fetcher will still show as `Bundler::Fetcher::Dependency`.

This means we will test the the gem's ruby requirement via this method: https://github.com/dependabot/dependabot-core/pull/2552/files#diff-3092b190ebf8948a50c8d770607f2417R172

My initial thought was that since the old method guards against the source being anything other than a `::Bundler::Source::Rubygems`, we should pass back the source as well and reinstate that check - which I think is a positive change - it doesn't fix this problem.

In the case where we are fetching from a private package manager the source will still be `::Bundler::Source::Rubygems` as proved by some of our gemfury specs elsewhere.

I think this means we have a pre-existing bug here where, if you have a private mirror of a gem that also exists in ruby gems, we will return update not possible if we find the rubygems version isn't compatible with your ruby version.

This feels like an extreme edge case and I'm reluctant to change existing behaviour as part of this diff but I just want to make sure my reasoning is correct.

Reply:
> My initial thought was that since the old method guards against the source being anything other than a ::Bundler::Source::Rubygems, we should pass back the source as well and reinstate that check - which I think is a positive change - it doesn't fix this problem.

The tests failed due to this check being missing - we can't get a fetcher from a non-rubygems source. This is fixed in 40a0464

Reply:
> I think this means we have a pre-existing bug here where, if you have a private mirror of a gem that also exists in ruby gems (or if a private gem shadows the name of a 'real' gem in ruby gems ), we will return update not possible if we find the rubygems version isn't compatible with your ruby version as we think we've retrieved it from the old rubygems API.
>
> This feels like an extreme edge case and I'm reluctant to change existing behaviour as part of this diff but I just want to make sure my reasoning is correct.

Good catch! Sounds plausible. Does indeed feel like an extreme edge case 😅  maybe worth adding a comment explaining this potential issue in case we run into it in future? 👍  on holding off on a fix in this PR.